tools/bug-report/pkg/bugreport/bugreport.go

```diff
@@ -0,0 +1,253 @@
+	// parsed with ParsePath.
+	logs       = make(map[string]string)
+	stats      = make(map[string]*processlog.Stats)
+	importance = make(map[string]int)
+	// Aggregated errors for all fetch operations.
+	gErrors util.Errors
+	lock    = sync.RWMutex{}
+)
+
+func runBugReportCommand(_ *cobra.Command) error {
+	// Default of "." doesn't work for K8s names.
+	util.PathSeparator = "/"
+
+	config, err := parseConfig()
+	if err != nil {
+		return err
+	}
+
+	_, clientset, err := kubeclient.New(config.KubeConfigPath, config.Context)
+	if err != nil {
+		return fmt.Errorf("could not initialize k8s client: %s ", err)
+	}
+	resources, err := cluster2.GetClusterResources(context.Background(), clientset)
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Cluster resource tree:\n\n%s\n\n", resources)
+	paths, err := filter.GetMatchingPaths(config, resources)
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Fetching logs for the following containers:\n\n%s\n", strings.Join(paths, "\n"))
+
+	gatherInfo(config, resources, paths)
+	if len(gErrors) != 0 {
+		log.Errora(gErrors.ToError())
+	}
+
+	// TODO: sort by importance and discard any over the size limit.
+	for path, text := range logs {
+		namespace, _, pod, _, err := cluster2.ParsePath(path)
+		if err != nil {
+			log.Errorf(err.Error())
+			continue
+		}
+		writeFile(archive.ProxyLogPath(tempDir, namespace, pod), text)
+	}
+	return nil
+}
+
+// gatherInfo fetches all logs, resources, debug etc. using goroutines.
+// proxy logs and info are saved in logs/stats/importance global maps.
+// Errors are reported through gErrors.
+func gatherInfo(config *config.BugReportConfig, resources *cluster2.Resources, paths []string) {
+	var wg sync.WaitGroup
+	cmdTimer := time.NewTimer(time.Duration(config.CommandTimeout))
+
+	getFromCluster(content.GetK8sResources, &content.Params{DryRun: config.DryRun}, archive.ClusterInfoPath(tempDir, "resources"), &wg)
+	getFromCluster(content.GetCRs, &content.Params{DryRun: config.DryRun}, archive.ClusterInfoPath(tempDir, "crs"), &wg)
+	getFromCluster(content.GetEvents, &content.Params{DryRun: config.DryRun}, archive.ClusterInfoPath(tempDir, "events"), &wg)
+	getFromCluster(content.GetClusterInfo, &content.Params{DryRun: config.DryRun}, archive.ClusterInfoPath(tempDir, "cluster-info"), &wg)
+	getFromCluster(content.GetSecrets, &content.Params{DryRun: config.DryRun, Verbose: config.FullSecrets}, archive.ClusterInfoPath(tempDir, "secrets"), &wg)
+	getFromCluster(content.GetDescribePods, &content.Params{DryRun: config.DryRun, Namespace: config.IstioNamespace}, archive.ClusterInfoPath(tempDir, "describe-pods"), &wg)
+
+	var wg2 sync.WaitGroup
+	for _, p := range paths {
+		p := p
+		namespace, _, pod, container, err := cluster2.ParsePath(p)
+		if err != nil {
+			log.Error(err.Error())
+			continue
+		}
+		wg2.Add(1)
+		go func() {
+			defer wg2.Done()
+			var wg3 sync.WaitGroup
```

Comment:
3 wait groups in one function seems like a code smell, probably very very hard to keep track.

Reply:
simplified this and named the wait groups more clearly.